bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableVaneerSettingsFactory.java

```diff
@@ -0,0 +1,312 @@
+    HeaderProvider headers = FixedHeaderProvider.create(GrpcUtil.USER_AGENT_KEY.name(), userAgent);
+
+    adminBuilder.stubSettings().setTransportChannelProvider(
+        InstantiatingGrpcChannelProvider.newBuilder()
+            .setHeaderProvider(headers)
+            .setEndpoint(endpoint)
+            .setPoolSize(options.getChannelCount())
+            .build());
+
+    return adminBuilder.build();
+  }
+
+  /**
+   * This method is use to build {@link BatchingSettings}.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildBulkMutationsSettings(Builder builder, BigtableOptions options) {
+    BulkOptions bulkOptions = options.getBulkOptions();
+    BatchingSettings.Builder batchSettingsBuilder = BatchingSettings.newBuilder();
+
+    long autoFlushMs = bulkOptions.getAutoflushMs();
+    long bulkMaxRowKeyCount = bulkOptions.getBulkMaxRowKeyCount();
+    long maxInflightRpcs = bulkOptions.getMaxInflightRpcs();
+    long shortRpcTimeoutMs = options.getCallOptionsConfig().getShortRpcTimeoutMs();
+
+    if (autoFlushMs > 0) {
+      batchSettingsBuilder.setDelayThreshold(ofMillis(autoFlushMs));
+    }
+    FlowControlSettings.Builder flowControlBuilder = FlowControlSettings.newBuilder();
+    if (maxInflightRpcs > 0) {
+      flowControlBuilder
+          .setMaxOutstandingRequestBytes(bulkOptions.getMaxMemory())
+          .setMaxOutstandingElementCount(maxInflightRpcs * bulkMaxRowKeyCount);
+    }
+
+    batchSettingsBuilder
+        .setIsEnabled(bulkOptions.useBulkApi())
+        .setElementCountThreshold(Long.valueOf(bulkOptions.getBulkMaxRowKeyCount()))
+        .setRequestByteThreshold(bulkOptions.getBulkMaxRequestSize())
+        .setFlowControlSettings(flowControlBuilder.build());
+
+    // TODO(rahulkql): implement bulkMutationThrottling & bulkMutationRpcTargetMs, once available
+    builder.bulkMutationsSettings()
+        .setBatchingSettings(batchSettingsBuilder.build())
+        .setSimpleTimeoutNoRetries(
+            ofMillis(options.getCallOptionsConfig().getShortRpcTimeoutMs()));
+  }
+
+  /**
+   * To build BigtableDataSettings#sampleRowKeysSettings with default retries settings.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildSampleRowKeysSettings(Builder builder, BigtableOptions options) {
+    builder.sampleRowKeysSettings()
+        .setRetrySettings(buildIdempotentRetrySettings(options));
+  }
+
+  /**
+   * To build BigtableDataSettings#mutateRowSettings with default retries settings.
+   *
+   * @param builder a {@link BigtableDataSettings.Builder} object.
+   * @param options a {@link BigtableOptions} object.
+   */
+  private static void buildMutateRowSettings(Builder builder, BigtableOptions options) {
+    builder.mutateRowSettings()
```

Comment:
where are the retry codes configured? (here and everywhere else)

Reply:
As of now, not overriding `RetryCodes` of `BigtableDataSettings`. 
Defaults retrycode in GCJ:
**[DEADLINE_EXCEEDED, UNAVAILABLE]** --> sampleRowKeysSettings, readRowsSettings, mutateRowSettings.
**empty** for bulkMutationsSettings, readModifyWriteRowSettings, checkAndMutateRowSettings.

Default retrycodes in [bigtable-client](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/RetryOptions.java#L50).

However, I can override these with operation specific retry codes.

Reply:
There are few more defaults for gcj in the overlay. The gapic generated client settings can't express all of the settings, so they are overlaid here:
https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java#L273-L321

This is where bulkMutations have their defaults set.

I'm trying to understand why we shouldn't copy the retry codes from cbt?

Reply:
I think for **Non-idempotent** operation we can utilize GCJ's default values, as we are overriding it with shortRpcTimeout & no retryCodes (_here both adapters have the default value of 1 min_).

In case of **Idempotent** operation default values varies alot & 
 - [CBT has retryCode](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/master/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/RetryOptions.java#L49-L54) as `{DEADLINE_EXCEEDED, UNAVAILABLE, ABORTED, UNAUTHENTICATED}` 
- [GCJ's retryCodes](https://github.com/googleapis/googleapis/blob/master/google/bigtable/v2/bigtable_gapic.yaml#L35-L39) are only `{DEADLINE_EXCEEDED, UNAVAILABLE}`.